Abstract Independence & Combined Abstract/Concrete POR

maul-esel opened the pull request:
This pull request merges GemCutter changes from the last few months back into `dev`.

## Abstract Independence
The PR  adds abstract independence relations (using variable-based abstraction by @lawki, and looper independence) for GemCutter. These independence relations allow more (or in general, *other*) statements to commute than the `SemanticIndependenceRelation` and `SyntacticIndependenceRelation` implemented so far, while still allowing for sound verification.

## New POR Algorithm
Furthermore, the PR adds a newly developed partial order reduction algorithm (called `SleepMapReduction`) that can take into account multiple independence relations at the same time. This still needs optimization work to be competitive, but the hope is that it will become GemCutter's new default reduction algorithm in the future.

## Overview of the Changes

* The changes in `trunk/examples` add a couple of new benchmarks and reorganize the GemCutter settings files.
* The changes in `Library-Automata` implement the new reduction algorithm, along with a few minor changes in `TotalizeNwa` and `ProductNwa`.
* In `Library-ModelCheckerUtils`, @Heizmann and I made some changes to support transferring `TransFormula`s with fresh term variables between scripts.
* In `Library-TraceCheckerUtils` we find the new independence relations and a lot of support infrastructure.
* `UltimateUtil` has some new code relating to lattices (needed for abstract independence refinement)
* `BasicCegarLoop` has a new feature useful for debugging: A set of SMT assertions can be read from a file and added to the proof, before the first iteration. This allows to e.g. check if the proof check actually succeeds on a proof that is expected to suffice for a given program.

***

I will do some more minor cleanup work in the next few days (mostly on the new benchmarks), and hope to merge this by Wednesday next week. Reviews are welcome :)

maul-esel committed:
add IRefinableIndependenceContainer

- better encapsulate abstraction level refinement
  (user of CEGAR loop need not know type of abstraction levels)
- will allow other kinds of abstract independence not induced
  by a refinable abstraction function

maul-esel committed:
reuse StateSplitter to decouple PartialOrderCegarLoop from reduction modes

maul-esel committed:
set a more reasonable timeout for ConcurrencyConfigurationQuickTest

maul-esel committed:
Merge branch 'wip/dk/abstract-por' into wip/dk/multi-por

maul-esel committed:
extract looper check as interface

maul-esel committed:
add LooperIndependenceRelation

maul-esel committed:
simplify UpsideDownLattice

lawki committed:
uptade for helperclass

lawki committed:
Merge branch 'wip/dk/abstract-por' of https://github.com/ultimate-pa/ultimate into wip/dk/abstract-por

lawki committed:
fnished helper class

lawki committed:
minor changes

lawki committed:
method refine specific implemented, maybe the method is bloated

lawki committed:
minor alignments. IDK really

lawki committed:
changed VarAbsConstraints. 6p.m. rathausplatz demonstration against war

lawki committed:
started with lattice

lawki committed:
chnages packaged of varAbConstraints

lawki committed:
Trying to revert wrnogly chnages MANIFEST.MF

maul-esel committed:
fix ComparisonResult::aggregate and expand documentation

maul-esel committed:
minor doc comment fixes

lawki committed:
changed project of VarAbConstraints

lawki committed:
rewritten Lattice

lawki committed:
fully implemented the lattice

lawki committed:
nothign really

lawki committed:
Beginning of new Class for specific Abstraction

lawki committed:
finished first try of complete SpecificVariableAbstraction

lawki committed:
changed the way variables are abstracted in SpecificAsbtarction

lawki committed:
aligned VariableAbstraction in regard to the new strategy of subsitutitng variables

lawki committed:
Build Path error after creating new TestClass for SpecificAbstraction

lawki committed:
still broken, fixed incompatabilities with immutable VarAbsConstraints-Class

maul-esel committed:
VariableAbstractionTest: wrap TF in IAction

maul-esel committed:
Merge branch 'dev' into wip/dk/abstract-por

maul-esel committed:
restore assert statement

maul-esel committed:
VariableAbstraction: add a new failing test

maul-esel committed:
fix failing auxVar test; some small cleanup

Different TFs must not share auxiliary variables. Hence,
we substitute all auxiliary variables by fresh copies.

maul-esel committed:
improved VariableAbstraction::restrict

maul-esel committed:
clean up dependencies

- switch SpecificVariableAbstractionTest to use BasicInternalAction
- undo unnecessary changes to MANIFEST.MF files
- undo accidental change to .classpath

lawki committed:
aligned tests of SpecificVariableAbstraction, added auxVar-copying, changed VarabLattice

lawki committed:
removed unnused method that was with errors in varAbConstraints

maul-esel committed:
new settings files for variable abstraction

maul-esel committed:
fix bug in SpecificVariableAbstraction

lawki committed:
deepcopy of maps where it is needed

maul-esel committed:
undo unintentional file changes

maul-esel committed:
restore VarAbsConstraints::hashCode

maul-esel committed:
make VarAbsConstraints truly immutable

maul-esel committed:
fix license for moved file

maul-esel committed:
VarAbsConstraints: remove unused methods

maul-esel committed:
refactor SpecificVariableAbstraction::restrict

- compute new constraints for letter in SpecificVariableAbstraction
- encapsulate creation of new constraint in VarAbsConstraints
- use a shallow map copy for time and memory efficiency

maul-esel committed:
simplify SpecificVariableAbstraction::refine

maul-esel committed:
restore assert in SpecificVariableAbstraction

maul-esel committed:
make auxiliary methods private

maul-esel committed:
cleanup obsolete code in PartialOrderCegarLoop

maul-esel committed:
use typical Ultimate naming

maul-esel committed:
VarAbLattice: remove commented code, remove useless map copies

maul-esel committed:
add support for top element to CanonicalLatticeForMaps

maul-esel committed:
make VarAbLattice nested class in VarAbsConstraints

This allows for better encapsulation:

- We no longer need to expose the maps, thus ensuring immutability.
- We can replace the internal data structures without affecting other classes.

maul-esel committed:
Merge branch 'dev' into wip/dk/abstract-por

# Conflicts:
#	trunk/source/Library-ModelCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/modelcheckerutils/cfg/transitions/UnmodifiableTransFormula.java
#	trunk/source/Library-TraceCheckerUtils/META-INF/MANIFEST.MF
#	trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/CegarLoopUtils.java
#	trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/TraceAbstractionObserver.java
#	trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/TraceAbstractionStarter.java

maul-esel committed:
move to proper package

maul-esel committed:
simplify tests with auxiliary methods

maul-esel committed:
add assert to catch bug

A state factory should not map multiple pairs of states to the same
product state, otherwise the product state's transitions can change
during runtime!

This assert fails when using PartialOrderCegarLoop.

maul-esel committed:
TotalizeNwa: support passing existing sink state

maul-esel committed:
PartialOrderCegarLoop: fix bug with ambiguous product state

If TotalizeNwa creates a new sink state with predicate TRUE, the state
factory used here does not distinguish between product states containing
the original TRUE predicate contained in the automaton and product states
containing the sink TRUE predicate.

This causes outgoing transitions of states to change, see e1cf1d48. As a
result, an old counterexample trace may no longer be accepted, and thus
be analyzed repeatedly.

We fix this by reusing the original TRUE as sink state.

maul-esel committed:
PartialOrderCegarLoop: properly support enhancement=NONE

The CEGAR loop behaved unsoundly, because it did not take into account
that the un-enhanced interpolant automata were nondeterministic.

Specifically, the TotalizeNwa pruned all outgoing edges when detecting
nondeterminism, which lead to feasible counterexamples being lost.
We fix this behaviour by explicitly determinizing the automata.

maul-esel committed:
TransFormulaUtils::checkImplication: support branch encoders

maul-esel committed:
comment assert for now

maul-esel committed:
ensure internal normal form

maul-esel committed:
check both transformulas' read variables

maul-esel committed:
VariableAbstraction: simply delegate to SpecificVariableAbstraction

This significantly reduces code duplication.

maul-esel committed:
Merge branch 'dev' into wip/dk/abstract-por

# Conflicts:
#	trunk/source/Library-ModelCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/modelcheckerutils/cfg/transitions/TransFormulaUtils.java

maul-esel committed:
fix block encoding regression introduced in 7a9d4c5

lawki committed:
removed unused code from testclasses of VariableAbstraction

maul-esel committed:
add tests to showcase bug in SpecificVariableAbstraction

maul-esel committed:
add read-only mode to dead end visitor

This can be helpful for debugging, and may have other applications in the
future: It allows us to use existing dead end information with visitors
that backtrack also on non-dead states, e.g. AutomatonConstructingVisitor.

maul-esel committed:
SpecificVariableAbstraction: avoid unnecessary HashSet copy

maul-esel committed:
fix debug message

maul-esel committed:
add SleepPredicate::toString

maul-esel committed:
add methods useful to debug unsoundness in PartialOrderCegarLoop

maul-esel committed:
simplify code using new helper method

maul-esel committed:
do not abstract unmodified variables that are partially constrained

If a variable that is not modified by a statement appears only
in the inConstraints, or only in the outConstraints, we treat it
as if it appeared in both.

The reason is that we can not abstract it properly without either

(1) breaking soundness wrt to valid Hoare triples, or
(2) making it an assigned variable of the abstracted statement.

For now (in this commit), we choose to preserve (2) [and of course, (1)]
and instead sacrifice the property that the abstracted statement read only
in-constrained variables.

maul-esel committed:
prepare evaluation

maul-esel committed:
move creation of abstraction function back into PartialOrderCegarLoop

- Immediate need to move it out no longer exists: Thanks to
  IRefinableIndependenceContainer, we no longer need H, the
  type parameter for the abstraction level.
- We need it to be in the same place as the creation of the
  independence relation, to use the independence script (to
  fix a bug occurring in the evaluation).
- Medium/long term, both aspects should be moved out again.

maul-esel committed:
introduce TransferrerWithVariableCache class

This class helps with transferring constructs involving IProgramVars
or IProgramConsts between scripts.

- modify TransFormulaBuilder::transferTransformula to use this class
- modify SemanticIndependenceRelation to use this class, and remove
  obsolete methods from TransFormulaBuilder
- add support to transfer IPredicates.
- pass transferrer to IndependenceBuilder, instead of creating it there

maul-esel committed:
IndependenceBuilder: add methods for conditional steps in fluent API

maul-esel committed:
simplify auxiliary functions in SpecificVariableAbstraction

maul-esel committed:
add support for transferring between scripts to SpecificVariableAbstraction

maul-esel committed:
fix locked ManagedScript in VariableAbstraction

CfgSmtToolkit's ManagedScript may be locked by Floyd/Hoare automaton,
in particular if enhancement is not NONE. This is a problem when the
SpecificVariableAbstraction e.g. declares new constants for auxVars:
An exception is thrown.

This bug was hidden by the fact that enabled assertions unlock the
ManagedScript before the exception occurs.

As a solution, we let the SpecificVariableAbstraction already transfer
TransFormulas to the script used for commutativity checking. We still
need to transfer conditions (in case of conditional independence).

If no abstraction is used, both TransFormulas and conditions are transferred
by the SemanticIndependenceRelation as before.

maul-esel committed:
use CfgSmtToolkit::createFreshManagedScript for independence script

This already transfers some symbols to the new script.
Axioms are also transferred.

maul-esel committed:
improve SpecificVariableAbstraction::refine

With the previous implementation, caching was severely limited:
As soon as the constraints for any letter was refined, the new
constraint was strictly less than the constrained returned by
previous calls to ::restrict, and the abstractions for all letters
were discarded.

The fixed implementation allows the constraints for other letters
to be refined all the way to the bottom-most abstraction level,
without invalidating the cached abstraction. Only if the constraints
for the cached letter are refined beyond a certain point, the cached
abstraction is invalidated.

maul-esel committed:
use test class for abstractions

Heizmann committed:
several workarounds to allow the transfer of Transformulas with fresh variables

maul-esel committed:
Merge branch 'dev' into wip/dk/abstract-por

# Conflicts:
#	trunk/source/Library-ModelCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/modelcheckerutils/cfg/transitions/UnmodifiableTransFormula.java
#	trunk/source/Library-TraceCheckerUtils/META-INF/MANIFEST.MF
#	trunk/source/Library-TraceCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/tracecheckerutils/partialorder/PartialOrderReductionFacade.java
#	trunk/source/Library-TraceCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/tracecheckerutils/partialorder/independence/abstraction/ICopyActionFactory.java
#	trunk/source/Library-TraceCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/tracecheckerutils/partialorder/independence/abstraction/IRefinableAbstraction.java
#	trunk/source/Library-TraceCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/tracecheckerutils/partialorder/independence/abstraction/RefinableCachedAbstraction.java
#	trunk/source/Library-TraceCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/tracecheckerutils/partialorder/independence/abstraction/SpecificVariableAbstraction.java
#	trunk/source/Library-TraceCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/tracecheckerutils/partialorder/independence/abstraction/VarAbsConstraints.java
#	trunk/source/Library-TraceCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/tracecheckerutils/partialorder/independence/abstraction/VariableAbstraction.java
#	trunk/source/Library-TraceCheckerUtilsTest/src/de/uni_freiburg/informatik/ultimate/lib/tracecheckerutils/partialorder/independence/abstraction/SpecificVariableAbstractionTest.java
#	trunk/source/Library-TraceCheckerUtilsTest/src/de/uni_freiburg/informatik/ultimate/lib/tracecheckerutils/partialorder/independence/abstraction/VariableAbstractionTest.java
#	trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/CegarLoopUtils.java
#	trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/TraceAbstractionStarter.java
#	trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/concurrency/PartialOrderCegarLoop.java
#	trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/preferences/TAPreferences.java
#	trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/preferences/TraceAbstractionPreferenceInitializer.java
#	trunk/source/UltimateTest/src/de/uni_freiburg/informatik/ultimate/ultimatetest/suites/traceabstraction/ConcurrencyConfigurationQuickTest.java

maul-esel committed:
prepare evaluation

maul-esel committed:
add missing settings files

maul-esel committed:
fix settings: disable Petri LBE

maul-esel committed:
Merge branch 'wip/dk/abstract-por' into wip/dk/multi-por

# Conflicts:
#	trunk/source/Library-TraceCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/tracecheckerutils/partialorder/independence/abstraction/ILooperCheck.java
#	trunk/source/Library-TraceCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/tracecheckerutils/partialorder/independence/abstraction/LooperIndependenceRelation.java
#	trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/concurrency/CegarLoopForPetriNet.java

maul-esel committed:
fix missing import

maul-esel committed:
pass copy factory directly to CegarLoopFactory constructor

maul-esel committed:
move abstraction type to IndependenceSettings

maul-esel committed:
rename method for clarity

maul-esel committed:
document new TotalizeNwa constructor

maul-esel committed:
some docs on independence settings

maul-esel committed:
add statistics support to IAbstraction

maul-esel committed:
document VarAbsConstraints, minor improvements

maul-esel committed:
reduce code duplication among abstraction tests

The common base class should also make it easier to add new tests,
particularly tests that apply to all abstraction implementations.

maul-esel committed:
re-enable assert in ProductNwa, fix AcceptingRunSearchVisitor to not trigger it

maul-esel committed:
fix settings, add settings check and TODO

maul-esel committed:
add auxiliary method for sanity checks

maul-esel committed:
re-disable assert statement once again

The fix in 7397240 did not fully remove this issue. It seems the only
way to fix this is either to allow this behaviour (remove the assert
permanently) or to remove the optimization in PartialOrderCegarLoop
that triggers it.

maul-esel committed:
add choice of solver and timeout for commutativity checks as settings

maul-esel committed:
move TF auxVar elimination to auxiliary method

maul-esel committed:
add quantifier elimination to variable abstraction

maul-esel committed:
dump SMT script for commutativity checks

maul-esel committed:
minor changes

maul-esel committed:
Merge branch 'dev' into wip/dk/abstract-por

maul-esel committed:
properly exit independence script

maul-esel committed:
fix preference declaration

maul-esel committed:
dump SMT in workdir (for now)

maul-esel committed:
fix simplification technique

maul-esel committed:
prepare evaluation

maul-esel committed:
Merge branch 'wip/dk/abstract-por' into wip/dk/multi-por

maul-esel committed:
export multireduction package

maul-esel committed:
begin implementing frontend support for multiple independence relations

maul-esel committed:
fix bug in sleep map update

maul-esel committed:
fix bug in DepthFirstTraversal

Always failed if visitor prunes all outgoing edges of starting state.
(For instance, if the starting state is a goal state.)

maul-esel committed:
use proper settings for abstraction type

maul-esel committed:
IndependenceRelationWithStatistics: record query statistics

maul-esel committed:
proper logging of independence settings

maul-esel committed:
fix typo in method name

maul-esel committed:
Add UnionPredicateCoverageChecker to combine coverage relations

maul-esel committed:
fix docs for ChainingHoareTripleChecker

maul-esel committed:
support LooperIndependenceRelation in PartialOrderCegarLoop

maul-esel committed:
update copyright

maul-esel committed:
improve statistics output

maul-esel committed:
prepare evaluation

maul-esel committed:
prepare evaluation

maul-esel committed:
prepare evaluation

maul-esel committed:
fix typo

maul-esel committed:
fix for automata without initial states

Our verification creates automata without initial states in certain
cases, e.g. when no thread can (locally) reach an error location.

Previously, this causes exceptions in the reduction and traversal
algorithms. Now, we output a warning and return gracefully.

maul-esel committed:
add a variant of CAMEL

maul-esel committed:
rename old example folder

maul-esel committed:
add some new examples

maul-esel committed:
prepare evaluation

maul-esel committed:
fix order for eval

maul-esel committed:
prepare evaluation

maul-esel committed:
sleep map: do not hash relations (causes nondeterminism)

maul-esel committed:
fix task set names: no spaces

maul-esel committed:
fix SleepMapReduction in case there are no initial states

Analogous to 7c7011cdc. Also, fix MinimalSleepSetReduction#isInitial.

maul-esel committed:
generalize example; add one more example program

maul-esel committed:
timeout is not an illegal state

maul-esel committed:
introduce IBudgetFunction interface

maul-esel committed:
refactor sleep set covering relation

maul-esel committed:
add a covering relation for automata from Petri nets

maul-esel committed:
simplify covering code

maul-esel committed:
extend documentation

maul-esel committed:
share independence cache

maul-esel committed:
implement data structure for subsets (using BitSet)

maul-esel committed:
use BitSubSet for VariableAbstraction abstraction levels

This avoids many expensive HashSet operations.
In one benchmark, this change lead to a speedup by a factor of 3.67!

maul-esel committed:
cache budget function

maul-esel committed:
cache SleepMap hashCode

maul-esel committed:
add a visitor for more light-weight reachability checks

maul-esel committed:
Improvements to OptimisticBudget

Enforce that callers use (possibly wrapped) AcceptingRunSearchVisitors.
This also allows us to get rid of the type parameter V.

Make sure to actually remember states from which an accepting state can
be reached, to avoid extremely expensive repeated recomputation of that
fact.

Also add some logging.

maul-esel committed:
some (questionable?) BitSubSet optimization

maul-esel committed:
SleepMapReduction: follow order in successor computation

Computing the successors requires the successor for smaller letters to
already be computed.  Hence, computing the successors out-of-order can
be expensive, when many of the computed successors are not needed.

maul-esel committed:
do not fail if no outgoing transition exists

maul-esel committed:
after timeout, do not start any new CEGAR loops

maul-esel committed:
prepare evaluation

maul-esel committed:
do not abstract TFs with branch encoders (not used)

maul-esel committed:
minor cleanup in DepthFirstTraversal

maul-esel committed:
more POPL2023 examples

maul-esel committed:
prepare evaluation

maul-esel committed:
add scaled versions of some examples where results differed for settings

maul-esel committed:
add variant of examples

maul-esel committed:
prepare evaluation

maul-esel committed:
xml fix

maul-esel committed:
fix xml

maul-esel committed:
refine persistent sets: join conflict

maul-esel committed:
prevent potential future bug in Petri2FiniteAutomatonAbstractionProvider

Hopeless cache must not be reused for different initial abstractions.
Hence, it must not be a field but should be a fresh map created for
each new call.

maul-esel committed:
add tailsum benchmarks

maul-esel committed:
one assert per program

maul-esel committed:
fix benchmark

maul-esel committed:
exclude some too difficult benchmarks

maul-esel committed:
add variant of benchmark

maul-esel committed:
support other budget functions

maul-esel committed:
add settings for coin flip budget

maul-esel committed:
small simplification

maul-esel committed:
add toString method for debugging purposes

maul-esel committed:
add missing file: CoinFLipBudget

maul-esel committed:
prepare evaluation

maul-esel committed:
fix setting validation

maul-esel committed:
switch to give probability in percent (integer)

maul-esel committed:
different seeds for different probabilities

maul-esel committed:
prepare evaluation

maul-esel committed:
variable renaming for clarity

maul-esel committed:
add more examples

maul-esel committed:
temporarily disable dead end optimization

maul-esel committed:
add setting for combo w/o conditional

maul-esel committed:
prepare evaluation

maul-esel committed:
add a number of new benchmarks

maul-esel committed:
Merge branch 'dev' into wip/dk/multi-por

maul-esel committed:
fix min/max examples

maul-esel committed:
add setting for coinflip seed

maul-esel committed:
prepare evaluation

maul-esel committed:
fix benchmark sets

maul-esel committed:
move example variants

maul-esel committed:
prepare evaluation

maul-esel committed:
add more examples & some proofs

maul-esel committed:
make sequential order more predictable: order by thread name

maul-esel committed:
do not share cache between abstract and concrete commutativity

In case abstract commutativity is syntactic-only, this unnecessarily
limits the power of concrete commutativity!

Medium-term we might want to find a way to share positive results
but not negative results. But the combination should remain predictable.

maul-esel committed:
eliminate quantifiers in abstraction even for syntactic independence

Doing so can lead to fewer "read" variables, allowing more syntactic
independence to be detected.

maul-esel committed:
implement (hackish) support for reading the proof from a file

maul-esel committed:
try coinflip budget without backtracking

maul-esel committed:
prepare evaluation

maul-esel committed:
disable reading proof

maul-esel committed:
prepare evaluation

maul-esel committed:
prepare evaluation

maul-esel committed:
add missing config files

maul-esel committed:
prepare eval

maul-esel committed:
fix xml

maul-esel committed:
prepare eval

maul-esel committed:
fix xml

maul-esel committed:
move duplicated helper method to one class

maul-esel committed:
move reading proofs to BasicCegarLoop, add setting

maul-esel committed:
DeterminizationFactory: ensure deterministic conjunction order

maul-esel committed:
add settings to control dumping of independence script

maul-esel committed:
externalize independence providers from POCegarLoop

maul-esel committed:
move GemCutter settings to own folder, remove some variants

maul-esel committed:
remove redundant benchexec file

maul-esel committed:
save statistics for all independence relations

maul-esel committed:
move classes to separate files

maul-esel committed:
cleaner linking of budget functions with reduction

maul-esel committed:
minor simplification

maul-esel committed:
fix unsoundness in Petri2AutomatonCoveringRelation

maul-esel committed:
some better comments

maul-esel committed:
record and re-use dead end information for only-concrete commutativity

schuessf reviewed (approved):
I am not really familiar with the partial order reduction implementation, but I support the general idea to combine multiple independence relations. At least on the parts I understand I could not spot any issue.

maul-esel committed:
move independence relation construction out of PartialOrderCegarLoop

maul-esel committed:
Merge branch 'dev' into wip/dk/multi-por

maul-esel committed:
Re-enable line (will evaluate effect)

maul-esel committed:
try fixing failing assert (disable optimization)

maul-esel committed:
prepare evaluation

maul-esel committed:
fix two off-by-one errors in BitSubSet

maul-esel committed:
don't even use DeadEndOptimizingSearchVisitor if not supported

There is no point in instantiating the visitor with a
dead end store that is always empty.

maul-esel committed:
add a kind of dead end support to ReachabilityCheckVisitor

This change reduced the verification time for the SV-COMP
benchmark pthread/triangular-1.i from ~190s to ~40s.

maul-esel committed:
explicitly declare settings for first three independence relations

This makes them e.g. accessible via command line and UI.
Also, improve the name of the setting

maul-esel committed:
some more documentation

maul-esel committed:
let BitSubSet not implement the Set interface

BitSubSet is meant for a more narrow use case.
E.g., not implementing Set<E> helps to ensure
that BitSubSets are not mixed with other kinds
of sets (eliminating the performance benefit).

This also eliminates the issue that BitSubSet
did not implement the ::hashCode contract of
Set<E>.

danieldietsch commented:
This will take some time, but I look into it this evening, perhaps I can at least skim ;)

maul-esel committed:
prepare evaluation

maul-esel committed:
(Specific)VariableAbstraction: overload constructors

maul-esel committed:
fix constant information for abstracted TFs

Through quantifier elimination and simplification, the abstracted TF
may contain fewer constant symbols than the original TF. This causes
an assert in UnmodifiableTransFormula::doConstantConsistencyCheck to
fail. Hence we filter the constants to those actually present.

maul-esel committed:
fix bug with already declared constants for auxiliary variables

This bug was reintroduced with commit 2175e79. Details are explained
in the comment in the code.

Instead of changing CfgSmtToolkit::createFreshManagedScript (used by
different callers), the code (now independent of the CfgSmtToolkit
anyway) is moved out.

maul-esel committed:
cleanup examples

maul-esel committed:
reduce code duplication

danieldietsch commented:
:)
